types/evidence.go

```diff
@@ -342,3 +355,581 @@ func (evl EvidenceList) Has(evidence Evidence) bool {
+	switch ev2.(type) {
+	case ConflictingHeadersEvidence:
+		e2 := ev2.(ConflictingHeadersEvidence)
+		return bytes.Equal(ev.H1.Hash(), e2.H1.Hash())
+	case *ConflictingHeadersEvidence:
+		e2 := ev2.(*ConflictingHeadersEvidence)
+		return bytes.Equal(ev.H1.Hash(), e2.H1.Hash())
+	default:
+		return false
+	}
+}
+
+func (ev ConflictingHeadersEvidence) ValidateBasic() error {
+	if ev.H1 == nil {
+		return errors.New("first header is missing")
+	}
+
+	if ev.H2 == nil {
+		return errors.New("second header is missing")
+	}
+
+	if err := ev.H1.ValidateBasic(ev.H1.ChainID); err != nil {
+		return fmt.Errorf("h1: %w", err)
+	}
+	if err := ev.H2.ValidateBasic(ev.H2.ChainID); err != nil {
+		return fmt.Errorf("h2: %w", err)
+	}
+	return nil
+}
+
+func (ev ConflictingHeadersEvidence) String() string {
+	return fmt.Sprintf("ConflictingHeadersEvidence{H1: %d#%X, H2: %d#%X}",
+		ev.H1.Height, ev.H1.Hash(),
+		ev.H2.Height, ev.H2.Hash())
+}
+
+//-------------------------------------------
+
+type PhantomValidatorEvidence struct {
+	Header *Header `json:"header"`
+	Vote   *Vote   `json:"vote"`
+}
+
+var _ Evidence = &PhantomValidatorEvidence{}
+var _ Evidence = PhantomValidatorEvidence{}
+
+func (e PhantomValidatorEvidence) Height() int64 {
+	return e.Header.Height
+}
+
+func (e PhantomValidatorEvidence) Time() time.Time {
+	return e.Header.Time
+}
+
+func (e PhantomValidatorEvidence) Address() []byte {
+	return e.Vote.ValidatorAddress
+}
+
+func (e PhantomValidatorEvidence) Hash() []byte {
+	bz := make([]byte, tmhash.Size+crypto.AddressSize)
+	copy(bz[:tmhash.Size-1], e.Header.Hash().Bytes())
+	copy(bz[tmhash.Size:], e.Vote.ValidatorAddress.Bytes())
+	return tmhash.Sum(bz)
+}
+
+func (e PhantomValidatorEvidence) Bytes() []byte {
+	return cdcEncode(e)
+}
+
+func (e PhantomValidatorEvidence) Verify(chainID string, pubKey crypto.PubKey) error {
+	// chainID must be the same
+	if chainID != e.Header.ChainID {
+		return fmt.Errorf("chainID do not match: %s vs %s",
+			chainID,
+			e.Header.ChainID,
+		)
+	}
+
+	if !pubKey.VerifyBytes(e.Vote.SignBytes(chainID), e.Vote.Signature) {
+		return errors.New("invalid signature")
+	}
+
+	return nil
+}
+
+func (e PhantomValidatorEvidence) Equal(ev Evidence) bool {
+	switch ev.(type) {
+	case PhantomValidatorEvidence:
+		e2 := ev.(PhantomValidatorEvidence)
+		return bytes.Equal(e.Header.Hash(), e2.Header.Hash()) &&
+			bytes.Equal(e.Vote.ValidatorAddress, e2.Vote.ValidatorAddress)
+	case *PhantomValidatorEvidence:
+		e2 := ev.(*PhantomValidatorEvidence)
+		return bytes.Equal(e.Header.Hash(), e2.Header.Hash()) &&
+			bytes.Equal(e.Vote.ValidatorAddress, e2.Vote.ValidatorAddress)
+	default:
+		return false
+	}
+}
+
+func (e PhantomValidatorEvidence) ValidateBasic() error {
+	if e.Header == nil {
+		return errors.New("empty header")
+	}
+
+	if e.Vote == nil {
+		return errors.New("empty vote")
+	}
+
+	// if err := e.Header.ValidateBasic(); err != nil {
+	// 	return fmt.Errorf("invalid header: %v", err)
+	// }
```

Comment:
don't you want to validate the header as well -> if not then remove the comments

Reply:
I'll uncomment once https://github.com/tendermint/tendermint/pull/4638 is merged